File: Settings/src/org/nameless/settings/fragment/PerAppSwitchConfigFragment.java
/*
 * Copyright (C) 2023-2024 The Nameless-AOSP Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.nameless.settings.fragment;

import android.content.Context;

import androidx.preference.Preference;
import androidx.preference.Preference.OnPreferenceChangeListener;

import java.util.ArrayList;
import java.util.List;
import java.util.LinkedHashMap;

import org.nameless.custom.preference.SwitchPreference;
import org.nameless.settings.widget.AppSwitchPreference;

public abstract class PerAppSwitchConfigFragment extends BasePerAppConfigFragment {

    private final LinkedHashMap<String, Boolean> mPkgCheckState = new LinkedHashMap<>();

    @Override
    protected Preference createAppPreference(Context prefContext, AppData appData) {
        final SwitchPreference pref = new AppSwitchPreference(prefContext);
        pref.setIcon(getIcon(appData.packageName));
        pref.setTitle(appData.label);
        pref.setSummary(appData.packageName);

        final boolean checked = isChecked(appData.packageName, appData.uid);
        pref.setChecked(checked);
        mPkgCheckState.put(appData.packageName, checked);

        pref.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object newValue) {
                final boolean ret = onSetChecked(appData.packageName, appData.uid, (Boolean) newValue);
                if (ret) {
                    mPkgCheckState.put(appData.packageName, (Boolean) newValue);
                    onCheckedListUpdated(generateCheckedPkgList());
                }
                return ret;
            }
        });

        return pref;
    }

    private List<String> generateCheckedPkgList() {
        final List<String> ret = new ArrayList<>();
        for (String pkg : mPkgCheckState.keySet()) {
            if (mPkgCheckState.get(pkg)) {
                ret.add(pkg);
            }
        }
        return ret;
    }

    protected abstract boolean isChecked(String packageName, int uid);

    protected abstract boolean onSetChecked(String packageName, int uid, boolean checked);

    protected abstract void onCheckedListUpdated(List<String> pkgList);
}

File: Settings/src/org/nameless/settings/fragment/PerAppSwitchConfigFragment.kt
/*
 * Copyright (C) 2025 The Nameless-CLO Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.nameless.settings.fragment

import android.content.Context

import androidx.preference.Preference

import org.nameless.custom.preference.SwitchPreferenceCompat
import org.nameless.settings.widget.AppSwitchPreference

abstract class PerAppSwitchConfigFragment : BasePerAppConfigFragment() {

    private val pkgCheckState = mutableMapOf<String, Boolean>()

    private var ignorePreferenceChange = false

    override fun createAppPreference(prefContext: Context, appData: AppData): Preference {
        return AppSwitchPreference(prefContext).apply {
            icon = getIcon(appData.packageName)
            title = appData.label
            summary = appData.packageName
            isChecked = isChecked(appData.packageName, appData.uid)

            pkgCheckState[appData.packageName] = isChecked

            setOnPreferenceChangeListener { p, v ->
                if (ignorePreferenceChange) {
                    return@setOnPreferenceChangeListener true
                }
                onSetChecked(
                    p as SwitchPreferenceCompat,
                    appData.packageName,
                    appData.uid,
                    v.toString().toBoolean()
                ).let {
                    if (it) {
                        pkgCheckState[appData.packageName] = v.toString().toBoolean()
                        onCheckedListUpdated(pkgCheckState.filter { it.value }.map { it.key })
                    }
                    return@setOnPreferenceChangeListener it
                }
            }
        }
    }

    override fun getSelectAllRunnable(): Runnable? {
        return Runnable {
            ignorePreferenceChange = true
            allPreference.forEach {
                (it as AppSwitchPreference).isChecked = true
            }
            pkgCheckState.mapValues { true }
            onCheckedListUpdated(pkgCheckState.filter { it.value }.map { it.key })
            ignorePreferenceChange = false
        }
    }

    override fun getDeselectAllRunnable(): Runnable? {
        return Runnable {
            ignorePreferenceChange = true
            allPreference.forEach {
                (it as AppSwitchPreference).isChecked = false
            }
            pkgCheckState.mapValues { false }
            onCheckedListUpdated(pkgCheckState.filter { it.value }.map { it.key })
            ignorePreferenceChange = false
        }
    }

    override fun getResetRunnable(): Runnable? {
        return Runnable {
            ignorePreferenceChange = true
            allPreference.forEach {
                (it as AppSwitchPreference).isChecked = getDefaultChecked()
            }
            pkgCheckState.mapValues { getDefaultChecked() }
            onCheckedListUpdated(pkgCheckState.filter { it.value }.map { it.key })
            ignorePreferenceChange = false
            showResetSuccessToast()
        }
    }

    open fun getDefaultChecked() = false

    abstract fun isChecked(packageName: String, uid: Int): Boolean

    abstract fun onSetChecked(pref: SwitchPreferenceCompat, packageName: String, uid: Int, checked: Boolean): Boolean

    abstract fun onCheckedListUpdated(pkgList: List<String>)
}
